### PR TITLE
handle GOTOOLCHAIN in kube::golang::verify_go_version

### PR DESCRIPTION
**Pull request by BenTheElder**
for now:
- shim FORCE_HOST_GO to default GOTOOLCHAIN=local (ensures we really force local go with 1.21+)
- treat a set GOTOOLCHAIN like FORCE_HOST_GO (so users can start idiomatically using the standard env in the future)
- set GOTOOLCHAIN to the selected go version otherwise (potentially unnecessary in the future, but it enforces that .go-version is the source of truth otherwise for now), keep gimme as a fallback
- pass GOTOOLCHAIN through to build containers

TODO (future PR): set toolchain statements in go.mod files and keep them in sync?

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

https://go.dev/doc/toolchain#GOTOOLCHAIN

#### Does this PR introduce a user-facing change?

```release-note
makefile and scripts now respect GOTOOLCHAIN and otherwise ensure ./.go-version is used
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

Holding for further discussion
/hold

### CODE REVIEW

**Comment by BenTheElder** [comment]
/sig release
/cc @liggitt @dims

Open question: Originally I implemented this to only treat `GOTOOLCHAIN=local` like `FORCE_HOST_GO=y`, since that's  the equivalent setting and not a default value. However now I think it might make more sense to treat "GOTOOLCHAIN is set at all"  to be like `FORCE_HOST_GO=y` and just defer to go's toolchain support. This would not make sense if we think users have errant GOTOOLCHAIN in their system env, which is unlikely unless they've done something like exported `go env` which will have `GOTOOLCHAIN=auto` by default.

Other than that debatable detail, I think this PR makes sense as a starting point to ensure no surprises when we go to switch to 1.21+

**Review by liggitt — COMMENTED** [review]
I think this makes sense, maybe wait for the go1.21 bump PR to merge first (https://github.com/kubernetes/kubernetes/pull/118996)?

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -467,12 +467,17 @@ kube::golang::verify_go_version() {
...
+  if [ -n "${FORCE_HOST_GO:-}" ]; then
+    GOTOOLCHAIN="${GOTOOLCHAIN:-local}"
+    export GOTOOLCHAIN
+  elif ! ([ -n "${GOTOOLCHAIN:-}" ] || \
       (command -v go >/dev/null && [ "$(go version | cut -d' ' -f3)" = "go${GO_VERSION}" ])); then
       export GIMME_ENV_PREFIX=${GIMME_ENV_PREFIX:-"${KUBE_OUTPUT}/.gimme/envs"}
       export GIMME_VERSION_PREFIX=${GIMME_VERSION_PREFIX:-"${KUBE_OUTPUT}/.gimme/versions"}
       # eval because the output of this is shell to set PATH etc.
       eval "$("${KUBE_ROOT}/third_party/gimme/gimme" "${GO_VERSION}")"
+      export GOTOOLCHAIN='go'"${GO_VERSION}"
```
[the remark this answers is not in the document]
Comment (liggitt) [reply hack/lib/golang.sh:480]: (I'm not 100% sure whether $GO_VERSION can contain a trailing newline and if that could break us here)

[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/lib/golang.sh:480]: It's equivalent, `'go'` is more explicit that the first part is a literal with no expansion but not a big deal here. You _have_ to write bash like this if you want to concat certain literals with expanded variables.

gimme doesn't like GO_VERSION with a newline either and we already "handle it" by the file not having no newline, but we could orthogonally consider trimming newline in the line that reads the file

[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/lib/golang.sh:480]: (you can see lack of newline in https://github.com/kubernetes/kubernetes/pull/118996/files#diff-69c69535f549f1e70ce037415e46a2894f04d4147c7929de6ef1db8f72a0a136R1)

[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/lib/golang.sh]: I updated the PR to use the "go${GO_VERSION}" form

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -467,12 +467,17 @@ kube::golang::verify_go_version() {
   # default GO_VERSION to content of .go-version
   GO_VERSION="${GO_VERSION:-"$(cat "${KUBE_ROOT}/.go-version")"}"
   # only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match GO_VERSION
-  if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
+  # and GOTOOLCHAIN is not set (the forward thinking idiomatic alternative FORCE_HOST_GO)
+  if [ -n "${FORCE_HOST_GO:-}" ]; then
+    GOTOOLCHAIN="${GOTOOLCHAIN:-local}"
+    export GOTOOLCHAIN
+  elif ! ([ -n "${GOTOOLCHAIN:-}" ] || \
```
Comment (liggitt) [inline hack/lib/golang.sh:474]: wondering if we'll encounter things explicitly setting GOTOOLCHAIN to its default ("GOTOOLCHAIN=auto") and if we should make this only skip our go version setup if a non-default GOTOOLCHAIN value is detected

Comment (BenTheElder) [reply hack/lib/golang.sh:474]: yeah, https://github.com/kubernetes/kubernetes/pull/120279#issuecomment-1699824195 raises that possibility.

+1 to make "auto" behave like unset. I'll implement that

Comment (liggitt) [reply hack/lib/golang.sh:474]: This is a nit, but mixing if / elif and `||` block chaining is hard to follow... is there a clearer way to see the cases we're handling.

Also, I think we're missing a case where 1) GOTOOLCHAIN was not set, 2) GO_VERSION was set, 3) _go_ already exists and matches GO_VERSION. Shouldn't we be setting GOTOOLCHAIN in that case?

```
if FORCE_HOST_GO:
  export GOTOOLCHAIN=local
elseif non-empty non-"auto" GOTOOLCHAIN:
  no-op, honor GOTOOLCHAIN
elsif go exists and is already at the right version:
  export GOTOOLCHAIN="go${GO_VERSION}" // missing case
else
  gimme "${GO_VERSION}"
  export GOTOOLCHAIN="go${GO_VERSION}" // missing case
```

Comment (BenTheElder) [reply hack/lib/golang.sh]: we should now treat `auto` like unset

File: hack/lib/golang.sh [hunk 3]
```diff
@@ -467,12 +467,17 @@ kube::golang::verify_go_version() {
   # default GO_VERSION to content of .go-version
   GO_VERSION="${GO_VERSION:-"$(cat "${KUBE_ROOT}/.go-version")"}"
   # only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match GO_VERSION
-  if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
+  # and GOTOOLCHAIN is auto or unset (the forward thinking idiomatic alternative FORCE_HOST_GO)
+  if [ -n "${FORCE_HOST_GO:-}" ]; then
+    GOTOOLCHAIN="${GOTOOLCHAIN:-local}"
+    export GOTOOLCHAIN
+  elif ! (([ -n "${GOTOOLCHAIN:-}" ] && [ "${GOTOOLCHAIN}" != 'auto' ]) || \
```
Comment (liggitt) [inline hack/lib/golang.sh:474]: would a single `"${GOTOOLCHAIN:-auto}" != 'auto'` be more readable?

Comment (BenTheElder) [reply hack/lib/golang.sh:474]: I think that's only slightly confusing if you're not aware that we also intend explicit `"auto"` to be handled the same as unset or set to `""` but we have a comment just above, and it's otherwise easier to read.

Comment (BenTheElder) [reply hack/lib/golang.sh:474]: done

File: hack/lib/golang.sh (hunk 2 above)
Comment (BenTheElder) [reply hack/lib/golang.sh]: /hold
I missed https://github.com/kubernetes/kubernetes/pull/120279#discussion_r1321893169 in the prior updates.
I agree this is worth refactoring.

**Comment by BenTheElder** [comment]
/retest
(CI pod pending timeout ...)

This was rewritten to:
1. If `GOTOOLCHAIN` has a non-auto value, respect it
2. If `FORCE_HOST_GO` is set, shim that to `GOTOOLCHAIN=local`
3. Otherwise, set `GOTOOLCHAIN` to `go${GO_VERSION}` (from `./.go-version` file) 
3b. If go isn't installed or that doesn't result in the right version because go is <1.21, fallback on gimme (edited) 


Someday we can drop 3b), and we already won't wind up running it if the host has 1.21+

File: hack/lib/golang.sh [hunk 4]
```diff
@@ -466,13 +466,23 @@ kube::golang::create_gopath_tree() {
...
+    GOTOOLCHAIN="go${GO_VERSION}"
+    export GOTOOLCHAIN
+    # if go is either not installed or too old to respect GOTOOLCHAIN then use gimme
+    if ! (command -v go >/dev/null && [ "$(go version | cut -d' ' -f3)" = "${GOTOOLCHAIN}" ]); then
       export GIMME_ENV_PREFIX=${GIMME_ENV_PREFIX:-"${KUBE_OUTPUT}/.gimme/envs"}
       export GIMME_VERSION_PREFIX=${GIMME_VERSION_PREFIX:-"${KUBE_OUTPUT}/.gimme/versions"}
       # eval because the output of this is shell to set PATH etc.
       eval "$("${KUBE_ROOT}/third_party/gimme/gimme" "${GO_VERSION}")"
+    fi
   fi
```
Comment (liggitt) [inline hack/lib/golang.sh:487]: hmm, also plumb into the container env construction?

```diff
diff --git a/build/common.sh b/build/common.sh
index 2935df35645..0001a2d5825 100755
--- a/build/common.sh
+++ b/build/common.sh
@@ -540,6 +540,7 @@ function kube::build::run_build_command_ex() {
     --env "KUBE_CGO_OVERRIDES=' ${KUBE_CGO_OVERRIDES[*]:-} '"
     --env "FORCE_HOST_GO=${FORCE_HOST_GO:-}"
     --env "GO_VERSION=${GO_VERSION:-}"
+    --env "GOTOOLCHAIN=${GOTOOLCHAIN:-}"
     --env "GOFLAGS=${GOFLAGS:-}"
     --env "GOGCFLAGS=${GOGCFLAGS:-}"
     --env "SOURCE_DATE_EPOCH=${SOURCE_DATE_EPOCH:-}"
```

Comment (BenTheElder) [reply hack/lib/golang.sh:487]: Good catch! done.

**Comment by liggitt** [comment]
/lgtm
/approve
can unhold when you're ready

**Comment by BenTheElder** [comment]
Pretty confident this has reasonable behavior now.

I think somewhere in N+M releases we should revisit:
1. Dropping gimme when go 1.21+ is widely packaged etc. (though ... we may still want a fallback for users without go on the host at all ...)
2. Nudging towards `GOTOOLCHAIN=local` over `FORCE_HOST_GO=y` and dropping the latter
3. Drop `.go-version` and just use `toolchain` go.mod statements.

But this should be a reasonable starting point.

File: hack/lib/golang.sh [hunk 5]
```diff
@@ -466,13 +466,23 @@ kube::golang::create_gopath_tree() {
...
+  if [ "${GOTOOLCHAIN:-auto}" != 'auto' ]; then
+    # no-op, just respect GOTOOLCHAIN
+    :
+  elif [ -n "${FORCE_HOST_GO:-}" ]; then
+    # ensure existing host version is used, like before GOTOOLCHAIN existed
+    export GOTOOLCHAIN='local'
+  else
+    # otherwise, we want to ensure the go version matches GO_VERSION
+    GOTOOLCHAIN="go${GO_VERSION}"
+    export GOTOOLCHAIN
```
Comment (liggitt) [inline hack/lib/golang.sh:478]: I think this should move inside the if block where we don't have the go version we want... I'm seeing behavior where setting `GOTOOLCHAIN=go1.21.1` explicitly is making `go1.21.1` try to redownload (maybe related to build modifiers like boringcrypto be enabled)

Comment (liggitt) [reply hack/lib/golang.sh:478]: `go env GOVERSION` reports `go1.21.1 X:boringcrypto` when boringcrypto is enabled. The `go version` check we're doing below ignores any experiments, so when we force GOTOOLCHAIN to exactly `go1.21.1`, we might be creating the mismatch.

Worth filing an issue for the go team to clarify how we should be using GOTOOLCHAIN in a situation like this.

Comment (BenTheElder) [reply hack/lib/golang.sh:478]: ACK, I suspect `export GOTOOLCHAIN=local` is the right answer when using a toolchain like this.
If not I think we could add a check for `X:` experiments early in the version check and assume that we should behave like `GOTOOLCHAIN=local` if an experimental toolchain is being used.

Comment (liggitt) [reply hack/lib/golang.sh:478]: > ACK, I suspect export GOTOOLCHAIN=local is the right answer when using a toolchain like this

I don't think so ... `GOTOOLCHAIN=local` won't do the right thing when we propagate into a build container with a different go version (like our kube-cross images)

Comment (BenTheElder) [reply hack/lib/golang.sh:478]: [we've reached out for clarification with the go team and will follow-up...]

Comment (BenTheElder) [reply hack/lib/golang.sh:478]: forgot to circle back here -- The expectation is that users will use GOEXPERIMENT which works with GOTOOLCHAIN (and this PR), GOTOOLCHAIN will download the standard toolchain but you can use the standard toolchain with GOEXPERIMENT instead of a modified toolchain, and that issue generally exists for 1.21+ in all projects.
